restore/restore.go

```diff
@@ -0,0 +1,341 @@
+
+	// If true, replace all data on the server with the local backup.
+	// Otherwise only restore the requested org/bucket, leaving other data untouched.
+	Full bool
+}
+
+type Services struct {
+	RestoreService influxdb.RestoreService
+	OrgService     influxdb.OrganizationService
+	BucketService  influxdb.BucketService
+}
+
+type restoreRunner struct {
+	Services
+
+	kvManifest     *influxdb.ManifestKVEntry
+	shardManifests map[uint64]*influxdb.ManifestEntry
+
+	tenantService influxdb.TenantService
+	metaClient    *meta.Client
+	log           *zap.Logger
+}
+
+func RunRestore(ctx context.Context, req Request, svcs Services, log *zap.Logger) error {
+	runner := restoreRunner{
+		Services: svcs,
+		log:      log,
+	}
+
+	if err := runner.loadManifests(req.Path); err != nil {
+		return err
+	}
+
+	if req.Full {
+		return runner.fullRestore(ctx, req)
+	}
+	return runner.partialRestore(ctx, req)
+}
+
+func (r *restoreRunner) loadManifests(path string) error {
+	// Read all manifest files from path, sort in descending time.
+	manifests, err := filepath.Glob(filepath.Join(path, "*.manifest"))
+	if err != nil {
+		return err
+	} else if len(manifests) == 0 {
+		return nil
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(manifests)))
+
+	r.shardManifests = make(map[uint64]*influxdb.ManifestEntry)
+	for _, filename := range manifests {
+		// Skip file if it is a directory.
+		if fi, err := os.Stat(filename); err != nil {
+			return err
+		} else if fi.IsDir() {
+			continue
+		}
+
+		// Read manifest file for backup.
+		var manifest influxdb.Manifest
+		if buf, err := ioutil.ReadFile(filename); err != nil {
+			return err
+		} else if err := json.Unmarshal(buf, &manifest); err != nil {
+			return fmt.Errorf("read manifest: %v", err)
+		}
+
+		// Save latest KV entry (first in the sorted slice).
+		if r.kvManifest == nil {
+			r.kvManifest = &manifest.KV
+		}
+
+		// Load most recent backup per shard.
+		for i := range manifest.Files {
+			sh := manifest.Files[i]
+			if _, err := os.Stat(filepath.Join(path, sh.FileName)); err != nil {
+				continue
+			}
+
+			entry := r.shardManifests[sh.ShardID]
+			if entry == nil || sh.LastModified.After(entry.LastModified) {
+				r.shardManifests[sh.ShardID] = &sh
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *restoreRunner) fullRestore(ctx context.Context, req Request) error {
+	if err := r.restoreKV(ctx, req.Path); err != nil {
+		return err
+	}
+
+	for _, m := range r.shardManifests {
+		if err := r.restoreShard(ctx, req.Path, m); err != nil {
+			return err
+		}
+	}
+
+	r.log.Info("Full restore complete", zap.String("path", req.Path))
+	return nil
+}
+
+func (r *restoreRunner) partialRestore(ctx context.Context, req Request) error {
+	// Open meta store so we can iterate over metadata.
+	kvStore := bolt.NewKVStore(r.log, filepath.Join(req.Path, r.kvManifest.FileName))
+	if err := kvStore.Open(ctx); err != nil {
+		return err
+	}
+	defer kvStore.Close()
+
+	r.tenantService = tenant.NewService(tenant.NewStore(kvStore))
```

Comment:
Is there any cleanup on the tenant.NewStore or tenantService() necessary?

Reply:
Not that I can see (neither has a `Close()` method)